synedrion/src/cggmp21/sigma/aff_g_star.rs

```diff
@@ -0,0 +1,338 @@
+
+                (ephemeral, commitment)
+            })
+            .unzip();
+
+        let mut reader = XofHasher::new_with_dst(HASH_TAG)
+            // commitments
+            .chain(&commitments)
+            // public parameters
+            .chain(public.pk0.as_wire())
+            .chain(public.pk1.as_wire())
+            .chain(&public.cap_c.to_wire())
+            .chain(&public.cap_d.to_wire())
+            .chain(&public.cap_y.to_wire())
+            .chain(&public.cap_x)
+            .chain(aux)
+            .finalize_to_reader();
+
+        // Non-interactive challenge
+        let e = BitVec::from_xof_reader(&mut reader, P::SECURITY_PARAMETER);
+
+        let elements = ephemerals
+            .into_iter()
+            .zip(e.bits())
+            .map(|(ephemeral, e_bit)| {
+                let z = if *e_bit {
+                    ephemeral.alpha + secret.x
+                } else {
+                    ephemeral.alpha
+                };
+                let z_prime = if *e_bit {
+                    // DEVIATION FROM THE PAPER.
+                    // See the comment in `AffGStarPublicInputs`.
+                    // Original: $+ y$. Modified: $- y$
+                    ephemeral.beta + secret.y.neg()
+                } else {
+                    ephemeral.beta
+                };
+
+                let exponent = if *e_bit {
+                    PublicSigned::one()
+                } else {
+                    PublicSigned::zero()
+                };
+                let w = secret.rho.to_masked(&ephemeral.r, &exponent);
+                // DEVIATION FROM THE PAPER.
+                // See the comment in `AffGStarPublicInputs`.
+                // Original: $\mu^{e_j}$. Modified: $\mu_{-e_j}$.
+                let lambda = secret.mu.to_masked(&ephemeral.s, &-exponent);
+
+                AffGStarProofElement {
+                    z: z.to_public(),
+                    z_prime: z_prime.to_public(),
+                    w,
+                    lambda,
+                }
+            })
+            .collect::<Vec<_>>();
+
+        Self {
+            e,
+            elements: elements.into(),
+            commitments: commitments.into(),
+        }
+    }
+
+    #[allow(clippy::too_many_arguments)]
+    pub fn verify(&self, public: AffGStarPublicInputs<'_, P>, aux: &impl Hashable) -> bool {
+        assert!(public.cap_c.public_key() == public.pk0);
+        assert!(public.cap_d.public_key() == public.pk0);
+        assert!(public.cap_y.public_key() == public.pk1);
+
+        let mut reader = XofHasher::new_with_dst(HASH_TAG)
+            // commitments
+            .chain(&self.commitments)
+            // public parameters
+            .chain(public.pk0.as_wire())
+            .chain(public.pk1.as_wire())
+            .chain(&public.cap_c.to_wire())
+            .chain(&public.cap_d.to_wire())
+            .chain(&public.cap_y.to_wire())
+            .chain(&public.cap_x)
```

Comment:
Do you really need to clone the booleans? Maybe `copied` is enough?

Reply:
It is. Resolves to the same assembly, of course, but I guess it's more idiomatic. Fixed.